src/apple/alohalytics_objc.mm

```diff
@@ -0,0 +1,354 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#if ! __has_feature(objc_arc)
+#error This file must be compiled with ARC. Either turn on ARC for the project or use -fobjc-arc flag
+#endif
+
+#import "../alohalytics_objc.h"
+#include "../alohalytics.h"
+#include "../logger.h"
+
+#include <utility> // std::pair
+#include <sys/xattr.h>
+#include <TargetConditionals.h> // TARGET_OS_IPHONE
+
+#import <CoreFoundation/CoreFoundation.h>
+#import <CoreFoundation/CFURL.h>
+#import <Foundation/NSURL.h>
+#import <UIKit/UIDevice.h>
+#import <UIKit/UIScreen.h>
+#import <UIKit/UIApplication.h>
+#import <AdSupport/ASIdentifierManager.h>
+
+using namespace alohalytics;
+
+namespace {
+// Conversion from [possible nil] NSString to std::string.
+static std::string ToStdString(NSString * nsString) {
+  if (nsString) {
+    return std::string([nsString UTF8String]);
+  }
+  return std::string();
+}
+
+// Additional check if object can be represented as a string.
+static std::string ToStdStringSafe(id object) {
+  if ([object isKindOfClass:[NSString class]]) {
+    return ToStdString(object);
+  } else if ([object isKindOfClass:[NSObject class]]) {
+    return ToStdString(((NSObject *)object).description);
+  }
+  return "ERROR: Trying to log neither NSString nor NSObject-inherited object.";
+}
+
+// Safe conversion from [possible nil] NSDictionary.
+static TStringMap ToStringMap(NSDictionary * nsDictionary) {
+  TStringMap map;
+  for (NSString * key in nsDictionary) {
+    map[ToStdString(key)] = ToStdStringSafe([nsDictionary objectForKey:key]);
+  }
+  return map;
+}
+
+// Safe conversion from [possible nil] NSArray.
+static TStringMap ToStringMap(NSArray * nsArray) {
+  TStringMap map;
+  std::string key;
+  for (id item in nsArray) {
+    if (key.empty()) {
+      key = ToStdStringSafe(item);
+      map[key] = "";
+    } else {
+      map[key] = ToStdStringSafe(item);
+      key.clear();
+    }
+  }
+  return map;
+}
+
+// Safe extraction from [possible nil] CLLocation to alohalytics::Location.
+static Location ExtractLocation(CLLocation * l) {
+  Location extracted;
+  if (l) {
+    if (l.horizontalAccuracy >= 0) {
+      extracted.SetLatLon([l.timestamp timeIntervalSince1970] * 1000.,
+                           l.coordinate.latitude, l.coordinate.longitude,
+                           l.horizontalAccuracy);
+    }
+    if (l.verticalAccuracy >= 0) {
```

Comment:
These conditions might benefit from a comment that `accuracy < 0 <=> location is unknown`.